logging/resource.go

```diff
@@ -0,0 +1,266 @@
+		},
+	}
+}
+
+func isCloudRun() bool {
+	_, config := os.LookupEnv("K_CONFIGURATION")
+	_, service := os.LookupEnv("K_SERVICE")
+	_, revision := os.LookupEnv("K_REVISION")
+	return config && service && revision
+}
+
+func detectCloudRunResource() *mrpb.MonitoredResource {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil
+	}
+	return &mrpb.MonitoredResource{
+		Type: "cloud_run_revision",
+		Labels: map[string]string{
+			"project_id":         projectID,
+			"location":           regionFromZone(zone),
+			"service_name":       os.Getenv("K_SERVICE"),
+			"revision_name":      os.Getenv("K_REVISION"),
+			"configuration_name": os.Getenv("K_CONFIGURATION"),
+		},
+	}
+}
+
+func isKubernetesEngine() bool {
+	clusterName, err := metadata.InstanceAttributeValue("cluster-name")
+	// Note: InstanceAttributeValue can return "", nil
+	if err != nil || clusterName == "" {
+		return false
+	}
+	return true
+}
+
+func detectKubernetesResource() *mrpb.MonitoredResource {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil
+	}
+	clusterName, err := metadata.InstanceAttributeValue("cluster-name")
+	if err != nil {
+		return nil
+	}
+	namespaceBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	namespaceName := ""
+	if err == nil {
+		namespaceName = string(namespaceBytes)
+	}
+	return &mrpb.MonitoredResource{
+		Type: "k8s_container",
+		Labels: map[string]string{
+			"cluster_name":   clusterName,
+			"location":       zone,
+			"project_id":     projectID,
+			"pod_name":       os.Getenv("HOSTNAME"),
+			"namespace_name": namespaceName,
+		},
+	}
+}
+
+func detectGCEResource() *mrpb.MonitoredResource {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil
+	}
+	id, err := metadata.InstanceID()
+	if err != nil {
+		return nil
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil
+	}
+	name, err := metadata.InstanceName()
+	if err != nil {
+		return nil
+	}
+	return &mrpb.MonitoredResource{
+		Type: "gce_instance",
+		Labels: map[string]string{
+			"project_id":    projectID,
+			"instance_id":   id,
+			"instance_name": name,
+			"zone":          zone,
+		},
+	}
+}
+
+func detectResource() *mrpb.MonitoredResource {
+	detectedResource.once.Do(func() {
+		switch {
+		// AppEngine, Functions, CloudRun are detected first, as metadata.OnGCE()
+		// erroneously returns true on these runtimes.
+		case isAppEngine():
+			detectedResource.pb = detectAppEngineResource()
+		case isCloudFunction():
+			detectedResource.pb = detectCloudFunction()
+		case isCloudRun():
+			detectedResource.pb = detectCloudRunResource()
+		case isKubernetesEngine():
+			detectedResource.pb = detectKubernetesResource()
+		case metadata.OnGCE():
+			detectedResource.pb = detectGCEResource()
+		}
+	})
+	return detectedResource.pb
```

Comment:
Shouldn't there be a default global resource here?

Reply:
this is handled by a separate function already see `monitoredResource`